### PR TITLE
Fix queue depth and related replica stats

### PR DESCRIPTION
**Pull request by KiranRG**
This diff contains two related changes:
1. Add/fix code in storehost to publish first/last message address and timestamps .. mostly intended for use for the queue-depth code
2. Fix queue-depth code and tests, that were breaking because of the fix above.

### CODE REVIEW

File: services/controllerhost/queueDepth.go [hunk 1]
```diff
@@ -329,9 +326,13 @@ func (qdc *queueDepthCalculator) computeBacklog(cgDesc *shared.ConsumerGroupDesc
...
 	switch qdc.iter.isDLQ {
 	case true:
-		backlog = common.MaxInt64(0, storeMetadata.lastSequence-(storeMetadata.beginSequence+1))
+		if storeMetadata.lastSequence != math.MaxInt64 && storeMetadata.beginSequence != math.MaxInt64 {
+			backlog = storeMetadata.lastSequence - storeMetadata.beginSequence + 1
+		}
 	case false:
-		backlog = common.MaxInt64(0, storeMetadata.availableSequence-cgExtent.GetAckLevelSeqNo())
+		if storeMetadata.availableSequence != math.MaxInt64 {
+			backlog = storeMetadata.availableSequence - cgExtent.GetAckLevelSeqNo()
```
Comment (venkat1109) [inline services/controllerhost/queueDepth.go]: you want to still leave the common.MaxInt64(0, backlog) as is - these values are eventually consistent.. So, there is no strong guarantee that AckLevelSeqNo will be less than the availableSequence. And to be consistent, you can do the same thing for if (true) case

File: services/controllerhost/queueDepthCache.go [hunk 2]
```diff
@@ -96,9 +95,6 @@ func (cache *storeExtentMetadataCache) fetchStoreExtentMetadata(extID string, st
 	}
 
 	rs := stats.GetReplicaStats()[0]
-	// store apparently reports illegal values
-	// fix them if needed
-	cache.fixReplicaStatsIfBroken(rs)
```
Comment (venkat1109) [inline services/controllerhost/queueDepthCache.go]: still want to leave this around to be backward compatible.. We should be able to delete this code once all the data has been touched by your new reporting.

File: services/storehost/extMgr.go [hunk 3]
```diff
@@ -783,39 +776,57 @@ func (ext *extentContext) delete() {
...
+func (ext *extentContext) initFirstMsg() error {
 
 	// FIXME: for timer-queues, this would return the seqNum of the first
 	// _available_ message to deliver, and *not* the first enqueued message.
 
 	_, key, err := ext.store.SeekFirst()
 
-	if err == nil && key != storage.InvalidKey {
-		_, beginSeqNum = ext.deconstructKey(key)
+	if err == nil && key != storage.InvalidKey &&
```
Comment (venkat1109) [inline services/storehost/extMgr.go]: positive assertions are more readable then negative assertions. In addition,
https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow

So, prefer

if err !=nil || key == storage.InvalidKey || ext.isSealExtentKey(key) {
     // init
    return
}

File: services/storehost/extMgr.go [hunk 4]
```diff
@@ -783,39 +776,57 @@ func (ext *extentContext) delete() {
...
 	// FIXME: for timer-queues, this would return the seqNum of the last
 	// _available_ message to deliver, and *not* the last enqueued message.
 
-	_, key, err := ext.store.SeekLast()
+	// find the last message (excluding any seal-key markers)
+	lastPossibleKey := ext.constructSealExtentKey(0) - 1
+
+	_, key, err := ext.store.SeekFloor(lastPossibleKey)
 
 	if err == nil && key != storage.InvalidKey {
```
Comment (venkat1109) [inline services/storehost/extMgr.go]: see comment above about control-flow

File: services/storehost/inConn.go [hunk 5]
```diff
@@ -303,7 +304,7 @@ func (t *inConn) writeMessagesPumpAppendOnly(msgC <-chan *inMessage, ackC chan<-
 			}
 
 			// if seqNum is not in expected (continuously increasing) order, log error as FYI
-			if x.getLastSeqNum()+1 != msgSeqNum && x.getLastSeqNum() != -1 {
+			if x.getLastSeqNum()+1 != msgSeqNum && x.getLastSeqNum() != math.MaxInt64 {
```
Comment (venkat1109) [inline services/storehost/inConn.go]: You might be better off defining a const for invalid first/last {seq,addr,ts} values. const helps if (1) ever you change this value to something else other than math.MaxInt64 (2) Finding out all references where invalid_ values are accessed in code.

File: services/controllerhost/queueDepth.go [hunk 6]
```diff
@@ -544,10 +545,17 @@ func (qdc *queueDepthCalculator) handleStartFrom(
 		if qualify {
 			trace += 100
 			consumerGroupExtent.WriteTime = common.Int64Ptr(int64(now))
-			consumerGroupExtent.AckLevelSeqNo = common.Int64Ptr(common.MaxInt64(
-				storeMetadata.beginSequence, // Retention may have removed some messages
-				int64(startFromSeq),         // Otherwise, act like we had just opened this extent at startFrom, i.e. don't count messages before startFrom
-			))
+
+			// retention may have purged some messages, account for that
+			if storeMetadata.beginSequence != math.MaxInt64 {
```
Comment (venkat1109) [inline services/controllerhost/queueDepth.go]: (1) this changes behavior, you still want to update AckLevelSeqNo if quality=true
(2) trace logging is broken
(3) Needs to be backwards compatible (i.e. also deal with beginSeq=-1)

File: services/controllerhost/queueDepth.go [hunk 7]
```diff
@@ -329,9 +326,13 @@ func (qdc *queueDepthCalculator) computeBacklog(cgDesc *shared.ConsumerGroupDesc
 
 	switch qdc.iter.isDLQ {
 	case true:
-		backlog = common.MaxInt64(0, storeMetadata.lastSequence-(storeMetadata.beginSequence+1))
+		if storeMetadata.lastSequence != math.MaxInt64 && storeMetadata.beginSequence != math.MaxInt64 {
+			backlog = storeMetadata.lastSequence - storeMetadata.beginSequence + 1
```
Comment (venkat1109) [inline services/controllerhost/queueDepth.go]: breaks backward compatibility

File: services/storehost/storehost.go [hunk 8]
```diff
@@ -789,21 +785,21 @@ func (t *StoreHost) SealExtent(ctx thrift.Context, req *store.SealExtentRequest)
 		// of 'seqNumUnspecifiedSeal' to go through, to support sealing an extent as a best effort.
 
 		switch {
-		case lastSeqNum == -1 && unspecifiedSealSeqNum: // no sealSeqNum specified and we don't know lastSeqNum
+		case lastSeqNum == math.MaxInt64 && unspecifiedSealSeqNum: // no sealSeqNum specified and we don't know lastSeqNum
```
Comment (venkat1109) [inline services/storehost/storehost.go]: this change isn't backward compatible

File: test/integration/integration_test.go [hunk 9]
```diff
@@ -2386,15 +2387,19 @@ func (s *NetIntegrationSuiteParallelB) TestQueueDepth() {
 				stats := srs.GetExtent().GetReplicaStats()[0] // Modify the existing stats. We would otherwise write nils to most fields
 
 				if reset { // clearing
-					if bs > 0 {
+					if bs != math.MaxInt64 {
```
Comment (venkat1109) [inline test/integration/integration_test.go]: compatibility

File: services/storehost/inConn.go (hunk 5 above)
Comment (KiranRG) [inline services/storehost/inConn.go]: Done.

File: services/controllerhost/queueDepth.go [hunk 10]
```diff
@@ -324,14 +325,20 @@ func (qdc *queueDepthCalculator) addExtentBacklog(
 
 func (qdc *queueDepthCalculator) computeBacklog(cgDesc *shared.ConsumerGroupDescription, cgExtent *metadata.ConsumerGroupExtent, storeMetadata *storeExtentMetadata, storeID string, logger bark.Logger) int64 {
 
-	var backlog int64
+	var backlog int64 = 0 // default
```
Comment (venkat1109) [inline services/controllerhost/queueDepth.go]: initialization is redundant. Defaults to zero anyways.

File: services/controllerhost/queueDepth.go [hunk 11]
```diff
@@ -542,12 +549,27 @@ func (qdc *queueDepthCalculator) handleStartFrom(
...
+
+			// retention may have purged some messages, account for that
+			if storeMetadata.beginSequence != math.MaxInt64 {
+
+				trace += 100
+
+				// the 'ack-level' corresponds to message that has already been read (ie 'acked');
+				// so reduce one from the 'beginSequence' to since that msg has not been read.
+				consumerGroupExtent.AckLevelSeqNo = common.Int64Ptr(common.MaxInt64(
+					storeMetadata.beginSequence-1, // Retention may have removed some messages
```
Comment (venkat1109) [inline services/controllerhost/queueDepth.go]: can the beginSequence be 0 ?

File: services/controllerhost/queueDepth.go [hunk 12]
```diff
@@ -324,14 +325,20 @@ func (qdc *queueDepthCalculator) addExtentBacklog(
...
-		backlog = common.MaxInt64(0, storeMetadata.lastSequence-(storeMetadata.beginSequence+1))
+		// update backog, only if the begin/first seqnums are available
+		if storeMetadata.lastSequence != math.MaxInt64 && storeMetadata.beginSequence != math.MaxInt64 {
+			backlog = storeMetadata.lastSequence - storeMetadata.beginSequence + 1
+		}
 	case false:
-		backlog = common.MaxInt64(0, storeMetadata.availableSequence-cgExtent.GetAckLevelSeqNo())
+		// update backlog, only if there is an available seqnum
+		if storeMetadata.availableSequence != math.MaxInt64 {
+			backlog = storeMetadata.availableSequence - cgExtent.GetAckLevelSeqNo()
```
Comment (venkat1109) [inline services/controllerhost/queueDepth.go]: I believe you still need max(0, diff) for backwards compatibility. If I understand right, the availableSequence can be -1 for empty extents with the old code.

File: services/controllerhost/queueDepth.go (hunk 11 above)
Comment (KiranRG) [inline services/controllerhost/queueDepth.go]: No. Message seq-nums start from "1". If there are no messages in the extent, then the beginSequence would be "MaxInt64".

File: services/controllerhost/queueDepth.go (hunk 12 above)
Comment (KiranRG) [inline services/controllerhost/queueDepth.go]: The 'fixReplicaStatsIfBroken' is expected to "fix" such values (those generated from the old behaviour) .. and make them "MaxInt64" (as is expected from the new behaviour).

File: services/controllerhost/queueDepth.go (hunk 10 above)
Comment (KiranRG) [inline services/controllerhost/queueDepth.go]: I put that to make it explicit that we rely on a default of "0". I'll probably just leave a comment there that says this.

**Pull request merged**